src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/Downloader.java

```diff
@@ -0,0 +1,49 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.bazel.repository.downloader;
+
+import com.google.common.base.Optional;
+import com.google.devtools.build.lib.events.ExtendedEventHandler;
+import com.google.devtools.build.lib.vfs.Path;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+
+public interface Downloader {
+  /**
+   * Downloads a file and writes it to the specified output path.
+   *
+   * <p>On failure the output file will be in an undefined state, and may
+   * or may not exist. The caller is responsible for cleaning up outputs of
+   * failed downloads.
+   *
+   * @param urls list of mirror URLs with identical content
+   * @param checksum valid checksum which is checked, or absent to disable
+   * @param output path to the destination file to write
+   * @throws IOException if download was attempted and ended up failing
+   * @throws InterruptedException if this thread is being cast into oblivion
+   */
+  void download(
+      List<URL> urls,
+      Map<URI, Map<String, String>> authHeaders,
+      Optional<Checksum> checksum,
+      String canonicalId,
```

Comment:
If I understand this change correctly, the downloader should manage the actual download but not care about the cache. So why do we keep the `canonicalId` part of the interface? (The only implementation of this interface does not use it.)

Reply:
Good catch. I fixed this in the internal review.

Reply:
The gRPC remote downloader client will need to pass this canonical ID through to the backend.